staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/compilation.go

```diff
@@ -92,30 +105,47 @@ func Compile(s *schema.Structural, isResourceRoot bool) ([]CompilationResult, er
 	// compResults is the return value which saves a list of compilation results in the same order as x-kubernetes-validations rules.
 	compResults := make([]CompilationResult, len(celRules))
 	for i, rule := range celRules {
-		var compilationResult CompilationResult
-		if len(strings.TrimSpace(rule.Rule)) == 0 {
-			// include a compilation result, but leave both program and error nil per documented return semantics of this
-			// function
-		} else {
-			ast, issues := env.Compile(rule.Rule)
-			if issues != nil {
-				compilationResult.Error = &Error{ErrorTypeInvalid, "compilation failed: " + issues.String()}
-			} else if !proto.Equal(ast.ResultType(), decls.Bool) {
-				compilationResult.Error = &Error{ErrorTypeInvalid, "cel expression must evaluate to a bool"}
-			} else {
-				prog, err := env.Program(ast)
-				if err != nil {
-					compilationResult.Error = &Error{ErrorTypeInvalid, "program instantiation failed: " + err.Error()}
-				} else {
-					compilationResult.Program = prog
-				}
+		compileRule(&compResults[i], rule, env)
+	}
+
+	return compResults, nil
+}
+
+func compileRule(compilationResult *CompilationResult, rule apiextensions.ValidationRule, env *cel.Env) {
+	if len(strings.TrimSpace(rule.Rule)) == 0 {
+		// include a compilation result, but leave both program and error nil per documented return semantics of this
+		// function
+		return
+	}
+	ast, issues := env.Compile(rule.Rule)
+	if issues != nil {
+		compilationResult.Error = &Error{ErrorTypeInvalid, "compilation failed: " + issues.String()}
+		return
+	}
+	if !proto.Equal(ast.ResultType(), decls.Bool) {
+		compilationResult.Error = &Error{ErrorTypeInvalid, "cel expression must evaluate to a bool"}
+		return
+	}
+
+	if checkedExpr, err := cel.AstToCheckedExpr(ast); err != nil {
+		// should be impossible since env.Compile returned no issues
+		compilationResult.Error = &Error{ErrorTypeInternal, "unexpected error: " + err.Error()}
+		return
+	} else {
+		for _, ref := range checkedExpr.ReferenceMap {
```

Comment:
would be good to get a double-check from Tristan that there's no way to reference `oldSelf` in an expression where it would not appear in this map

Reply:
@TristonianJones Could you take a quick look at this, please? Is this a correct/reasonable way to determine whether or not an expression contains a reference to a specific identifier?

Reply:
That looks right to me. If `oldSelf` is not referenced then there will not be a `Reference` to it in the `checkedExpr.ReferenceMap`